File: 0611. Valid Triangle Number.java
// Medium
// Two Pointers
// Time:O(n^2), Space:O(1)
// https://leetcode.cn/problems/valid-triangle-number/

import java.util.*;

// a+b>c a<=b<=c

class Solution {
    public int triangleNumber(int[] nums) {
        int cnt = 0;
        Arrays.sort(nums);
        
        for (int k = nums.length - 1; k >= 2; k--) {
            int left = 0, right = k - 1;

            while (left < right) {
                if (nums[left] + nums[right] > nums[k]) {
                    // [left+1...right-1]+right are fine
                    cnt += right - left;
                    right--;
                } else {
                    left++;
                }
            }
        }
        return cnt;
    }
}

File: 0501-1000/0611. Valid Triangle Number.py
# Time:O(n²), Space:O(1)

from typing import List

class Solution:
    def triangleNumber(self, nums: List[int]) -> int:
        nums.sort()
        n = len(nums)
        ans = 0

        # fix a side(largest one)
        for k in range(n - 1, -1, -1):
            left, right = 0, k - 1

            while left < right:
                if nums[left] + nums[right] > nums[k]:
                    ans += right - left
                    right -= 1
                else:
                    left += 1
        
        return ans
